test/op_shape_test.cpp

```diff
@@ -2283,6 +2283,198 @@ TEST_CASE(test_scalar_nelemnts)
+        std::vector<size_t> b{2};
+        migraphx::shape ds{dtype, b};
+
+        throws_shape(migraphx::make_op("gathernd"), is, ds);
+    }
+
+    {
+        // See Example 4 at https://github.com/onnx/onnx/blob/main/docs/Operators.md#GatherND
+        auto dtype = migraphx::shape::float_type;
+        auto itype = migraphx::shape::int64_type;
+        migraphx::shape is{itype, {2, 2}};
+        migraphx::shape ds{dtype, {2, 2}};
+
+        migraphx::shape s0{dtype, {2}};
+        expect_shape(s0, migraphx::make_op("gathernd"), ds, is);
+    }
+
+    {
+        // See Example 5 at https://github.com/onnx/onnx/blob/main/docs/Operators.md#GatherND
+        auto dtype = migraphx::shape::float_type;
+        auto itype = migraphx::shape::int64_type;
+        migraphx::shape is{itype, {2, 1}};
+        migraphx::shape ds{dtype, {2, 2, 2}};
+
+        int batch_dims(1);
+        migraphx::shape s0{dtype, {2, 2}};
+        expect_shape(s0, migraphx::make_op("gathernd", {{"batch_dims", batch_dims}}), ds, is);
+    }
+}
+
+TEST_CASE(test_gathernd_dynamic0)
+{
+    {
+        // k > r
+        auto dtype = migraphx::shape::float_type;
+        auto itype = migraphx::shape::int64_type;
+        migraphx::shape is{itype, {2, 4}};
+        std::vector<migraphx::shape::dynamic_dimension> b{{8, 8, 0}};
+        migraphx::shape ds{dtype, b};
+
+        int batch_dims(1);
+        throws_shape(migraphx::make_op("gathernd", {{"batch_dims", batch_dims}}), ds, is);
+    }
+}
+
+TEST_CASE(test_gathernd_dynamic1)
+{
+    {
+        // k > r - batch_dims
+        auto dtype = migraphx::shape::float_type;
+        auto itype = migraphx::shape::int64_type;
+        migraphx::shape is{itype, {2, 4}};
+        std::vector<migraphx::shape::dynamic_dimension> b{{2, 2, 0}};
+        migraphx::shape ds{dtype, b};
+
+        int batch_dims(1);
+        throws_shape(migraphx::make_op("gathernd", {{"batch_dims", batch_dims}}), ds, is);
+    }
+}
+
+TEST_CASE(test_gathernd_dynamic2)
+{
+    {
+        // int(q) + r - k - batch_dims - 1 = 0
+        auto dtype = migraphx::shape::float_type;
+        auto itype = migraphx::shape::int64_type;
+        migraphx::shape is{itype, {2, 4}};
+        std::vector<migraphx::shape::dynamic_dimension> b{{2, 2, 0}};
+        migraphx::shape ds{dtype, b};
+
+        throws_shape(migraphx::make_op("gathernd"), is, ds);
+    }
+}
+
+TEST_CASE(test_gathernd_dynamic3)
+{
+    {
+        // See Example 4 at https://github.com/onnx/onnx/blob/main/docs/Operators.md#GatherND
+        auto dtype = migraphx::shape::float_type;
+        auto itype = migraphx::shape::int64_type;
+        migraphx::shape is{itype, {2, 2}};
+        std::vector<migraphx::shape::dynamic_dimension> b{{2, 2, 0}, {2, 2, 0}};
+        migraphx::shape ds{dtype, b};
+
+        migraphx::shape::dynamic_dimension ddout{2, 2, 0};
+        migraphx::shape s0{dtype, {ddout}};
+        expect_shape(s0, migraphx::make_op("gathernd"), ds, is);
+    }
+}
```

Comment:
This is Example 1 from the ONNX spec, not Example 4. The `dynamic_dimension` is also fixed. We wouldn't see this case because the `dynamic_dimension` would be simplified during parsing to a static shape.

Reply:
I'll update the comment.  

Even if we wouldn't see the case, don't we still want to test the internal behavior?  Given that it's a unit and not end-to-end test.  Do we have, or need, tests of the conversion to static?

Reply:
Also, if I change all the fixed dynamic shapes I've got in my tests to be non-fixed, does it matter for purposes of testing whether I specify an opt value; set it to 0; or leave it blank?

Reply:
You specify the opt value so that you're testing whatever code you added for calculating it. Ex: if you coded it that the opt for the output shape should always be 0, test that with different inputs.